cairo/ethereum/cancun/vm/instructions/environment.cairo

```diff
@@ -439,3 +460,272 @@ func codecopy{range_check_ptr: felt, evm: Evm}() -> ExceptionalHalt* {
+    let ok = cast(0, ExceptionalHalt*);
+    return ok;
+}
+
+// @notice Returns the keccak256 hash of a contract's bytecode
+func extcodehash{
+    range_check_ptr,
+    poseidon_ptr: PoseidonBuiltin*,
+    bitwise_ptr: BitwiseBuiltin*,
+    keccak_ptr: KeccakBuiltin*,
+    evm: Evm,
+}() -> ExceptionalHalt* {
+    alloc_locals;
+
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        let (address_u256, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+    }
+
+    // GAS
+    let accessed_addresses = evm.value.accessed_addresses;
+    let accessed_addresses_ptr = cast(accessed_addresses.value.dict_ptr, DictAccess*);
+    tempvar address_u256_ = UnionUintU256(new UnionUintU256Enum(cast(0, Uint*), address_u256));
+    let address_ = to_address(address_u256_);
+    tempvar address = new Address(address_.value);
+    let (is_present) = hashdict_read{dict_ptr=accessed_addresses_ptr}(1, &address.value);
+    if (is_present == 0) {
+        // If the entry is not in the accessed storage keys, add it
+        hashdict_write{dict_ptr=accessed_addresses_ptr}(1, &address.value, 1);
+        tempvar poseidon_ptr = poseidon_ptr;
+        tempvar accessed_addresses_ptr = accessed_addresses_ptr;
+    } else {
+        tempvar poseidon_ptr = poseidon_ptr;
+        tempvar accessed_addresses_ptr = accessed_addresses_ptr;
+    }
+    let poseidon_ptr = cast([ap - 2], PoseidonBuiltin*);
+    let new_accessed_addresses_ptr = cast([ap - 1], SetAddressDictAccess*);
+    tempvar new_accessed_addresses = SetAddress(
+        new SetAddressStruct(
+            evm.value.accessed_addresses.value.dict_ptr_start, new_accessed_addresses_ptr
+        ),
+    );
+    EvmImpl.set_accessed_addresses(new_accessed_addresses);
+
+    let access_gas_cost = (is_present * GasConstants.GAS_WARM_ACCESS) + (1 - is_present) *
+        GasConstants.GAS_COLD_ACCOUNT_ACCESS;
+    let err = charge_gas(Uint(access_gas_cost));
+    if (cast(err, felt) != 0) {
+        return err;
+    }
+
+    // OPERATION
+    // Get the account from state
+    let state = evm.value.env.value.state;
+    let account = get_account{state=state}([address]);
+    let env = evm.value.env;
+    EnvImpl.set_state{env=env}(state);
+    EvmImpl.set_env(env);
+
+    let _empty_account = EMPTY_ACCOUNT();
+    let empty_account = OptionalAccount(_empty_account.value);
+    let is_empty_account = Account__eq__(OptionalAccount(account.value), empty_account);
+
+    // If account is empty, push 0
+    if (is_empty_account.value != 0) {
+        tempvar code_hash_u256 = U256(new U256Struct(0, 0));
+
+        tempvar keccak_ptr = keccak_ptr;
+        tempvar poseidon_ptr = poseidon_ptr;
+        tempvar range_check_ptr = range_check_ptr;
+        tempvar bitwise_ptr = bitwise_ptr;
+    } else {
+        // Calculate keccak256 hash of code and push to stack
+        let code_hash = keccak256(account.value.code);
```

Comment:
to debug: inner call to keccak fails in rust vm

Reply:
This fails here 
https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/cairo/common/builtin_keccak/keccak.cairo#L180-L185

Reply:
[04deb70](https://github.com/kkrt-labs/keth/pull/499/commits/04deb7092b44250a050bb50773c38a26e90aa28c)  ->  follow up issue https://github.com/kkrt-labs/keth/issues/502